_posts/2022-05-31-Alarmeboard_mit_Jira_und_Cloudwatch.md

```diff
@@ -0,0 +1,369 @@
+Das gängige Mittel der Wahl unter AWS sind Cloudwatch-Alarme, die Bedingungen an Metriken oder Protokollfilter knüpfen können.
+Sobald die entsprechende Bedingung erfüllt ist, wird eine Aktion ausgeführt.
+Dies kann zum Beispiel eine Skalierungsaktion für den entsprechenden Service oder auch das Senden eines Events z.B. an ein Topic im Simple Notification Service sein.
+Landen die Alarm-Events einmal in einem Topic ist es ein leichtes die Information via E-Mail an alle Entwickler zu verteilen, die darauf reagieren sollten.
+
+# Alarme mit Terraform erzeugen und an ein Topic schicken
+
+Als Beispiel soll uns hier ein Alarm dienen, der immer dann ein Event an ein konfigurierbares Topic sendet, wenn eine Message in einer Dead Letter Queue (kurz DLQ) gelandet ist:
+```terraform
+resource "aws_cloudwatch_metric_alarm" "dlq_alarm" {
+   alarm_name                = "MyProject-DlqAlarm-Stage"
+   comparison_operator       = "GreaterThanOrEqualToThreshold"
+   evaluation_periods        = 1
+   metric_name               = "ApproximateNumberOfMessagesVisible"
+   namespace                 = "AWS/SQS"
+   period                    = "60"
+   statistic                 = "Sum"
+   threshold                 = "1"
+   ok_actions                = []
+   insufficient_data_actions = []
+   datapoints_to_alarm       = 1
+   alarm_description         = "Sends an event when there are messages in the DLQ"
+   dimensions = {
+      "QueueName" = aws_sqs_queue.dlq.name
+   }
+   alarm_actions = [var.alarmTopicSns]
+   tags          = var.tags
+}
+```
+Damit könnte man meinen das Thema sei durch.
+Alle Entwickler, die mit ihrer E-Mail das Topic abonniert haben, wissen Bescheid.
+Ziel erreicht!
+
+Es ergeben sich aber tatsächlich mehrere Probleme, für die Abhilfe geschaffen werden kann:
+1. Wer bearbeitet das Problem eigentlich konkret?
+2. Wurden überhaupt alle Alarme bereits bearbeitet?
+3. Wie lesbar und damit wie schnell bearbeitbar sind diese erzeugten E-Mails überhaupt?
+
+Tatsache ist, dass wir hier auf ein Problem gestoßen sind.
+Es gibt immer wieder Alarme, die mehrfach auslösen, sodass es dazu kommen kann, dass mehrere Entwickler am gleichen Alarm arbeiten, ohne das kurzfristig zu erkennen.
+Gibt es mal eine größere Anzahl von Alarmen, ist das E-Mail-Postfach nicht der Ort, an dem man übersichtlich sicherstellen kann, dass alle Alarme bereits bearbeitet wurden.
+Und auch die Lesbarkeit der Mails lässt zu wünschen übrig.
+Es sind zwar alle notwendigen technischen Informationen enthalten, aber es beschleunigt die Bearbeitung ungemein, wenn man zum Beispiel einen Stacktrace aus den Logs direkt im Anhang findet.
+
+# Lambdas für die Erzeugung der Jira-Tickets nutzen
+
+> ### Outline
+> Jira bietet eine [REST-Schnittstelle](https://developer.atlassian.com/server/jira/platform/rest-apis/) an, die dafür genutzt werden kann, Alarme in Tickets zu gießen.
+> * Motivation für 2 Lambdas
```

Comment:
Motivation für 2 Lambdas fehlt noch. Vielleicht einfach eine Kurzübersicht über die Lösung bevor die Lambdas im Detail erklärt werden?

Reply:
Gar nicht so einfach. Ich versuche mal folgendes:

Wir teilen die Aufgaben für die Erzeugung und Anreicherung der Jira-Tickets in zwei Lambdas auf.
Damit erhöhen wir sowohl die Flexibilität als auch die Ausfallsicherheit der Lösung.
Sobald die erste Lambda für das initiale Anlegen eines Jira-Tickets erfolgreich durchgelaufen ist, können wir dieses Ticket um Zusatzinformationen, deren Beschaffung dann auch langläufiger oder komplizierter sein kann, erweitern.
In unserem Fall erweitern wir die Tickets bei protokollbasierten Alarmen im zweiten Schritt um den auslösenden Log-Eintrag.
Der Ansatz an sich ist aber offen für Erweiterungen, die aus den Anforderungen anderer Projekte entstehen könnten.